From 8ed9f8f36ca23a29db0849767d5f482ae07c1b1c Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Thu, 23 Jan 2025 11:45:36 +0800
Subject: [PATCH 1/5] AOSCOS: LOONGSON3: 50drm: limit DRM module installation
 to amdgpu, loongson, and radeon

If we are using mips64el (assumed to be MIPS-based Loongson), only install
modules for loongson, radeon, and amdgpu as these are the only three DRM
drivers supported by these platforms. We need to minimise the sizes of the
initramfs so that it fits within the RAM allocated by the system firmware.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 modules.d/50drm/module-setup.sh | 23 ++++++++++++++++++-----
 1 file changed, 18 insertions(+), 5 deletions(-)

diff --git a/modules.d/50drm/module-setup.sh b/modules.d/50drm/module-setup.sh
index 8e52f111..46e029eb 100755
--- a/modules.d/50drm/module-setup.sh
+++ b/modules.d/50drm/module-setup.sh
@@ -18,7 +18,11 @@ installkernel() {
             "=drivers/video/backlight"
     fi
 
-    instmods amdkfd hyperv_fb "=drivers/pwm"
+    if [[ ${DRACUT_ARCH:-$(uname -m)} != mips64 ]]; then
+        instmods amdkfd hyperv_fb "=drivers/pwm"
+    else
+        instmods "=drivers/pwm"
+    fi
 
     # if the hardware is present, include module even if it is not currently loaded,
     # as we could e.g. be in the installer; nokmsboot boot parameter will disable
@@ -50,9 +54,18 @@ installkernel() {
             instmods "$modname"
         done
     else
-        dracut_instmods -o -s "drm_crtc_init|drm_dev_register|drm_encoder_init" "=drivers/gpu/drm" "=drivers/staging"
-        # also include privacy screen providers (see above comment)
-        # atm all providers live under drivers/platform/x86
-        dracut_instmods -o -s "drm_privacy_screen_register" "=drivers/platform/x86"
+        if [[ ${DRACUT_ARCH:-$(uname -m)} != mips64 ]]; then
+            dracut_instmods -o -s "drm_crtc_init|drm_dev_register|drm_encoder_init" "=drivers/gpu/drm" "=drivers/staging"
+            # also include privacy screen providers (see above comment)
+            # atm all providers live under drivers/platform/x86
+            dracut_instmods -o -s "drm_privacy_screen_register" "=drivers/platform/x86"
+        else
+            # If we are using mips64 (assumed to be MIPS-based
+            # Loongson), only install loongson, radeon, and amdgpu as
+            # these are the only three DRM drivers supported by these
+            # platforms. We need to minimise the sizes of the initramfs so
+            # that it fits within the RAM allocated by the system firmware.
+            dracut_instmods -o -s "drm_crtc_init|drm_dev_register|drm_encoder_init" "=drivers/gpu/drm/amd/amdgpu" "=drivers/gpu/drm/amd/amdxcp" "=drivers/gpu/drm/loongson" "=drivers/gpu/drm/radeon"
+        fi
     fi
 }
-- 
2.48.1

